### ISSUE TITLE
BUG: Unexpected resulting shape after tuple indexing

### ISSUE THREAD

**Issue by class255** [numpy/numpy#23101]
### Describe the issue:

I found that tuple indexing for numpy.ndarray somtimes produces unexpected resulting shapes.

For example, when `mat` has shape `(4, 5, 6, 7)`,
`mat[..., (1,2), :, 0]` including move the axis of (1,2) (size 5) to the first, while `mat[..., (1,2), 0, 0]`, `mat[..., (1,2), 0, :]`, and `mat[..., (1,2), :, :]` maintain the position of the axis of (1,2). Replacing `...` to `:` yields the same results, and `mat[0, :, (1,2), ...]` behaves similarly.
For detail, please see the code block below (results are written as multi line comments).

These results are highly unintuitive and difficult to expect the resulting shapes produced by tuple indexing. In contrast, PyTorch's `torch.Tensor` behaves as expected, as shown in the last part of the code below.

Were these results intended? Then how can we expect resulting shapes properly when we use tuple indexing, single integer indexing, and `:` for different axes simultaneously?

### Reproduce the code example:

```python
import numpy as np
import torch

def shape_report(mat):
    print(f"{type(mat) = }")
    print(f"{mat[..., (1,2), 0, 0].shape = }")
    print(f"{mat[..., (1,2), 0, :].shape = }")
    print(f"{mat[..., (1,2), :, 0].shape = }")
    print(f"{mat[..., (1,2), :, :].shape = }")

    print(f"{mat[0, 0, (1,2), ...].shape = }")
    print(f"{mat[0, :, (1,2), ...].shape = }")
    print(f"{mat[:, 0, (1,2), ...].shape = }")
    print(f"{mat[:, :, (1,2), ...].shape = }")

mat = np.arange(4*5*6*7).reshape(4,5,6,7)
shape_report(mat)
''' Results:
type(mat) = <class 'numpy.ndarray'>
mat[..., (1,2), 0, 0].shape = (4, 2)
mat[..., (1,2), 0, :].shape = (4, 2, 7)
mat[..., (1,2), :, 0].shape = (2, 4, 6)
mat[..., (1,2), :, :].shape = (4, 2, 6, 7)
mat[0, 0, (1,2), ...].shape = (2, 7)
mat[0, :, (1,2), ...].shape = (2, 5, 7)
mat[:, 0, (1,2), ...].shape = (4, 2, 7)
mat[:, :, (1,2), ...].shape = (4, 5, 2, 7)
'''

print(f"{mat[:, (1,2), 0, 0].shape = }")
print(f"{mat[:, (1,2), 0, :].shape = }")
print(f"{mat[:, (1,2), :, 0].shape = }")
print(f"{mat[:, (1,2), :, :].shape = }")
''' Results:
mat[:, (1,2), 0, 0].shape = (4, 2)
mat[:, (1,2), 0, :].shape = (4, 2, 7)
mat[:, (1,2), :, 0].shape = (2, 4, 6)
mat[:, (1,2), :, :].shape = (4, 2, 6, 7)
'''

tensor = torch.arange(4*5*6*7).reshape(4,5,6,7)
shape_report(tensor)
''' Results:
type(mat) = <class 'torch.Tensor'>
mat[..., (1,2), 0, 0].shape = torch.Size([4, 2])
mat[..., (1,2), 0, :].shape = torch.Size([4, 2, 7])
mat[..., (1,2), :, 0].shape = torch.Size([4, 2, 6])
mat[..., (1,2), :, :].shape = torch.Size([4, 2, 6, 7])
mat[0, 0, (1,2), ...].shape = torch.Size([2, 7])
mat[0, :, (1,2), ...].shape = torch.Size([5, 2, 7])
mat[:, 0, (1,2), ...].shape = torch.Size([4, 2, 7])
mat[:, :, (1,2), ...].shape = torch.Size([4, 5, 2, 7])
'''
```

### Runtime information:

3.10.8 (tags/v3.10.8:aaaf517, Oct 11 2022, 16:50:30) [MSC v.1933 64 bit (AMD64)]
1.23.4

**Comment by rkern** [comment]
The behavior of mixing slices and advanced indexing is [documented here](https://numpy.org/devdocs/user/basics.indexing.html#combining-advanced-and-basic-indexing). There is a [proposal for other, more useful behaviors in NEP 21](https://numpy.org/neps/nep-0021-advanced-indexing.html).
